x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java

```diff
@@ -0,0 +1,260 @@
+        }
+    });
+
+    public SamlAuthnRequestValidator(SamlIdentityProvider idp) {
+        SamlUtils.initialize();
+        this.idp = idp;
+    }
+
+    public SamlValidateAuthnRequestResponse processQueryString(String queryString) {
+        final List<String> parameterList = Arrays.asList(queryString.split("&"));
+        final Map<String, String> parameters = new HashMap<>();
+        if (parameterList.isEmpty()) {
+            throw new IllegalArgumentException("Invalid Authentication Request query string");
+        }
+        RestUtils.decodeQueryString(queryString, 0, parameters);
+        logger.trace(new ParameterizedMessage("Parsed the following parameters from the query string: {}", parameters));
+        // We either expect at least a single parameter named SAMLRequest or at least 3 ( SAMLRequest, SigAlg, Signature )
+        final String samlRequest = parameters.get("SAMLRequest");
+        final String relayState = parameters.get("RelayState");
+        final String sigAlg = parameters.get("SigAlg");
+        final String signature = parameters.get("Signature");
+        if (null == samlRequest) {
+            logAndThrow(() -> new ParameterizedMessage("Query string [{}] does not contain a SAMLRequest parameter", queryString));
+        }
+        AuthnRequest authnRequest = null;
+        // We consciously parse the AuthnRequest before we validate its signature as we need to get the Issuer, in order to
+        // verify if we know of this SP and get its credentials for signature verification
+        final Element root = parseSamlMessage(inflate(decodeBase64(samlRequest)));
+        if ("AuthnRequest".equals(root.getLocalName()) == false
+            || "urn:oasis:names:tc:SAML:2.0:protocol".equals(root.getNamespaceURI()) == false) {
+            logAndThrow(() -> new ParameterizedMessage("SAML message [{}] is not an AuthnRequest", SamlUtils.toString(root, true)));
+        }
+        try {
+            authnRequest = SamlUtils.buildXmlObject(root, AuthnRequest.class);
+        } catch (Exception e) {
+            logAndThrow(() -> new ParameterizedMessage("Cannot process AuthnRequest [{}]", SamlUtils.toString(root, true)), e);
+        }
+        final SamlServiceProvider sp = getSpFromIssuer(authnRequest.getIssuer());
+        if (signature != null) {
+            if (sigAlg == null) {
+                logAndThrow(() ->
+                    new ParameterizedMessage("Query string [{}] contains a Signature but SigAlg parameter is missing",
+                        queryString));
+            }
+            final X509Credential spSigningCredential = sp.getSigningCredential();
+            if (spSigningCredential == null) {
+                logAndThrow(
+                    "Unable to validate signature of authentication request, Service Provider hasn't registered signing credentials");
+            }
+            validateSignature(samlRequest, sigAlg, signature, sp.getSigningCredential(), relayState);
+        }
```

Comment:
So we validate signatures, but we never care if a request is unsigned?
I don't know what we want (should an SP have a config for what sorts of messages they will sign, and we enforce signatures there?), but perhaps we at least need a `TODO` here.

Reply:
This can go a few ways: 
- We can demand _only_ signed AuthnRequests, declare that in our metadata too `WantAuthnRequestsSigned` and fail here. 
- Allow SPs to sign and handle this but not mandate it ( what is what we have so far )
- Set this in configuration when an SP is registered and act accordingly. 

I'll add a TODO so that we don't miss this and we can discuss

Reply:
Based on our discussion today, I went with the following: 

- If the SP should sign requests, this should be explicitly defined during registration
- If the SP should sign requests, and the request has a signature, we validate it
- If the SP should sign requests, and the request doesn't have a signature, we fail
- If the SP should not sign requests, we do not attempt to parse and validate a signature even if one exists